File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/net/java/truevfs/kernel/impl/ReentrantReadWriteLockAspect.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl

import java.util.concurrent.locks._
import javax.annotation.concurrent._

/** A mixin which provides some features of its associated reentrant read/write
  * `lock`.
  *
  * @author Christian Schlichtherle
  */
@ThreadSafe
private trait ReentrantReadWriteLockAspect
extends ReadWriteLockAspect[ReentrantReadWriteLock] {

  /** Returns `true` if and only if the read lock is held by the
    * current thread.
    * This method should only get used for assertions, not for lock control.
    * 
    * @return `true` if and only if the read lock is held by the
    *         current thread.
    */
  final def readLockedByCurrentThread: Boolean = 0 != lock.getReadHoldCount

  /** Returns `true` if and only if the write lock is held by the current
    * thread.
    * This method should only get used for assertions, not for lock control.
    * 
    * @return `true` if and only if the write lock is held by the current
    *         thread.
    */
   final def writeLockedByCurrentThread: Boolean = lock.isWriteLockedByCurrentThread

  /** Checks that the write lock is held by the current thread.
    * Use this method for lock control.
    * 
    * @throws NeedsWriteLockException if the `writeLock` is not held by the
    *         current thread.
    * @see    #writeLockedByCurrentThread
    */
  final def checkWriteLockedByCurrentThread(): Unit = {
    if (!writeLockedByCurrentThread) throw NeedsWriteLockException()
  }
}

File: src/main/java/net/java/truevfs/kernel/impl/ReentrantReadWriteLockAspect.java
/*
 * Copyright © 2005 - 2021 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl;

import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 * A mixin which provides some features of its reentrant read/write {@link #getLock()}.
 *
 * @author Christian Schlichtherle
 */
interface ReentrantReadWriteLockAspect extends ReadWriteLockAspect<ReentrantReadWriteLock> {

    /**
     * Returns {@code true} if and only if the read lock is held by the current thread.
     * This method should only get used for assertions, not for lock control.
     *
     * @return {@code true} if and only if the read lock is held by the current thread.
     */
    default boolean readLockedByCurrentThread() {
        return 0 != getLock().getReadHoldCount();
    }

    /**
     * Returns {@code true} if and only if the write lock is held by the current thread.
     * This method should only get used for assertions, not for lock control.
     *
     * @return {@code true} if and only if the write lock is held by the current thread.
     */
    default boolean writeLockedByCurrentThread() {
        return getLock().isWriteLockedByCurrentThread();
    }

    /**
     * Checks that the write lock is held by the current thread.
     * Use this method for lock control.
     *
     * @throws NeedsWriteLockException if the {@link #writeLock()} is not held by the current thread.
     * @see #writeLockedByCurrentThread
     */
    default void checkWriteLockedByCurrentThread() {
        if (!writeLockedByCurrentThread()) {
            throw NeedsWriteLockException.apply();
        }
    }
}
